man/man1/extract.1
.TH extract 1 "19 May 1998"
.SH NAME
extract \- extract data from a cube
.SH SYNOPSIS
.PP
.B extract -i infile [-o outfile] [mode]
.PP
.B extract --in infile [--out outfile] [mode]
.SH DESCRIPTION
.PP
.B extract
is used to extract data from a cube.
Several modes of extraction are implemented : plane extraction,
cube extraction (continuously from one plane to another), pattern
extraction, quadrant extraction, list of planes, and rectangle (slit)
extraction.
.PP
Details are given below on modes and syntax.
.SH COMMAND LINE
.B extract
receives always at least one option: the input file name, provided by
\-i or \--in. Providing the output file name by \-o or \--out is not
mandatory and sometimes not possible. Defaults for output names depend
on the extraction mode. 
.SH EXTRACTION MODES
.TP
.BI "\-m " plane " -b " num " -e " num 
extracts planes to single files. Plane numbers go from
1 to the number of planes in the cube (NAXIS3). Default value for
beginning plane is 1, default value for end plane is NAXIS3.
To extract one plane only, give the same number for begin and end.
.TP
.BI "\-m " cube " -b " num " -e " num
extracts a cube from another, edge planes are included.
Default value for begin plane is 1, default value for end plane
is NAXIS3.
.TP
.BI "\-m " quad " -x " 1234
extracts quadrants from a cube. Quadrants to extract
shall be requested by a character string containing
the quadrant numbers.
Quadrants are named according to the following convention:
.br
1 2
.br
4 3
.TP
.BI "\-m " pattern " -x " "01101001... " "\-s " step 
extracts planes in a cube according to a pattern.
A pattern is a character string containing only
0 and 1's, planes identified with a 1 are extracted,
those with a 0 are not. The pattern is periodically
repeated.
.PP
The step option identifies how many planes are taken into account per
cycle step. e.g. an extraction pattern of 01, with a step parameter
of 5 would yield an actual extraction pattern of 0000011111.
This parameter defaults to 1 plane per cycle step.
.TP
.BI "\-m " list " -f " file
This option uses a list of plane numbers given into a text file.
It extracts the requested planes to one cube.
.PP
It is mainly intended to be used with
.B stcube
which would first extract all statistical information from a cube,
then use a selector program to extract plane numbers corresponding
to planes which have interesting statistical values, and then
extract only these planes to a cube. 
.TP
.BI \-m " rect " -x " 'LLX LLY URX URY'"
This option extracts a rectangular area over the whole cube.
Input option defines the area by feeding in the lower left corner
and upper right corner coordinates. The rectangle itself limiting
the area is included in the extracted part.  Do not forget to input them 
within simple quotes! The coordinate system is given according to
the FITS reference:
.PP
First pixel in image is in lower left corner, and has coordinates
(1,1). Last pixel in image is in upper right corner of the image
and has coordinates (NAXIS1, NAXIS2).
.PP
LLX and LLY are the coordinates of lower left corner.
.br
URX and URY are the coordinates of upper right corner.
.SH LONG OPTIONS
.PP
Here are long option equivalents for all command-line switches:
.PP
General options:
.br
\-i         = \--in
.br
\-o         = \--out
.PP
Mode names:
.br
\-m plane   = \--plane
.br
\-m cube    = \--cube
.br
\-m quad    = \--quadrant
.br
\-m pattern = \--pattern
.br
\-m list    = \--list
.br
\-m rect    = \--rectangle
.PP
Mode options:
.br
\-b         = \--begin
.br
\-e         = \--end
.br
\-x         = \--ext
.br
\-f         = \--name
.SH FILES
.PP
Input files shall all comply with FITS format
Output files have the same pixel type as input files.
.SH NAMING SCHEMES
In single file output, default name is 'out.fits'. In multiple
file output, default names depend on the extraction mode:
.PP
in plane mode : basename.0001.fits to basename.9999.fits
.PP
in quadrant mode : basename.quad.1.fits to basename.quad.4.fits
.PP
The original FITS header of the input FITS file (-i option) 
is conserved along, except for the following keywords:
NAXIS, NAXISn, BITPIX, BSCALE, BZERO, which are related to the
newly created file.
.PP
HISTORY keywords are appended to the FITS header to indicate the
eclipse process modifications.
.SH EXAMPLES
.PP
To extract plane 53 out of cube 'incube.fits' (output name is optional, 
default would be incube.0053.fits):
.PP
> extract -m plane -i incube.fits -b 53 -e 53 -o p53.fits
.PP
To extract all planes of file 'incube.fits' to different FITS files 
(multiple file output, output name option is not considered):
.PP
> extract -m plane -i incube.fits
.br
would generate files names such as incube.0001.fits and so on, to last
plane number.
.PP
To extract all planes from plane 10 to last one in 'incube.fits', into a 
single cube named 'truncated.fits':
.PP
> extract -m cube -i incube.fits -b 10 -o truncated.fits
.PP
To extract only quadrants 2 and 4 from 'incube.fits':
(multiple file output, output name option is not considered):
.PP
> extract -m quad -x 24 -i incube.fits
.br
Output files are named 'incube.quad.2.fits' and 'incube.quad.4.fits'.
.PP
To extract only even planes from 'incube.fits' to 'even.fits':
.PP
> extract -m pattern -x '01' -i incube.fits -o even.fits
.PP
To extract planes 2, 3, 5, 7 and 11 in cube incube.fits to one cube
named partial.fits, create a file named listp for example, and type
in the plane numbers:
.PP
---- beginning of file 'listp' ----
.br
2
.br
3
.br
5
.br
7
.br
---- end of file 'listp' -----
.PP
The command is then:
.br
> extract -m list -f listp -i incube.fits -o partial.fits
.PP
To extract both lower quadrants in a 256x256 image named in.fits,
into an image named lower.fits:
.br
> extract -m rect -x '1 1 256 128' -i in.fits -o lower.fits
.PP
The following commands are equivalent:
.br
> extract -m pattern -x '0000011111' -i infile.fits -o out.fits
.br
> extract -m pattern -x '01' -s 5 -i infile.fits -o out.fits
.SH SEE ALSO
.PP
.B catcube, stcube
.SH BUGS
.PP
Misuse of command line switches may result in undefined behaviour.


CHANGES
List of changes between eclipse 4.9 and eclipse 5.0
---------------------------------------------------

Since last delivery, there has been no major channge in the library.
Only a couple of minor fixes were brought to the ISAAC pipeline.



src/main/extract.c
/*----------------------------------------------------------------------------*/
/**
   @file    extract.c
   @author  Nicolas Devillard
   @date    October 13, 1995
   @version	$Revision: 1.23 $
   @brief   Extract data from a cube
*/
/*----------------------------------------------------------------------------*/

/*
	$Id: extract.c,v 1.23 2002/11/19 15:27:19 yjung Exp $
	$Author: yjung $
	$Date: 2002/11/19 15:27:19 $
	$Revision: 1.23 $
 */

/*-----------------------------------------------------------------------------
   								Includes
 -----------------------------------------------------------------------------*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "eclipse.h"

/*-----------------------------------------------------------------------------
   								Defines
 -----------------------------------------------------------------------------*/

#define OPT_PLANE		1001
#define OPT_CUBE		1002
#define	OPT_QUADRANT	1003
#define OPT_PATTERN		1004
#define OPT_LIST		1005
#define OPT_RECTANGLE	1006

#define OPT_BEGIN		2001
#define OPT_END			2002
#define OPT_EXTSTRING	2003
#define OPT_LISTNAME	2004
#define OPT_STEP		2005

/*-----------------------------------------------------------------------------
   								New types
 -----------------------------------------------------------------------------*/

typedef enum extraction_mode {
	EXT_PLANE,
	EXT_PATTERN,
	EXT_CUBE,
	EXT_QUAD,
	EXT_LIST,
	EXT_RECTANGLE,
	EXT_UNDEFINED
} extraction_mode ;

static int cube_extract(
	char			*	name_i,
	char			*	name_o,
	extraction_mode		emode,
	int					begin,
	int					end,
	char			*	listname,
	char			*	extstring,
	int					step
) ;

/*-----------------------------------------------------------------------------
   								Function prototypes
 -----------------------------------------------------------------------------*/

static char prog_desc[] = "extract data from a cube" ;
static void usage(char * pname) ;

/*-----------------------------------------------------------------------------
  								    Main
 -----------------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
	char		    name_i[FILENAMESZ+1],
				    name_o[FILENAMESZ+1],
				    listname[FILENAMESZ+1] ;
	char		    extstring[MAX_IMAGE_NUMBER] ;
	int			    begin, 
                    end ;
	int			    step ;
	extraction_mode emode ;
	int			    c ;

	if (argc<2) usage(argv[0]) ;

    /* Initialize */
	name_i[0] = name_o[0] = listname[0] = 0 ; 
	begin = end = -1 ;
	emode = EXT_UNDEFINED ;
	memset(extstring, 0, MAX_IMAGE_NUMBER);
	step = 1 ;

    while (1) {
        int     option_index = 0 ;
        static struct option long_options[] =
        {
            {"license", 0, 0, OPT_LICENSE},
            {"help",    0, 0, OPT_HELP},
            {"version", 0, 0, OPT_VERSION},

            {"plane",     0, 0, OPT_PLANE},
            {"cube",      0, 0, OPT_CUBE},
            {"quadrant",  0, 0, OPT_QUADRANT},
            {"pattern",   0, 0, OPT_PATTERN},
            {"list",      0, 0, OPT_LIST},
            {"rectangle", 0, 0, OPT_RECTANGLE},

            {"begin", 1, 0, OPT_BEGIN},
            {"end",   1, 0, OPT_END},
            {"ext",   1, 0, OPT_EXTSTRING},
            {"name",  1, 0, OPT_LISTNAME},
            {"step",  1, 0, OPT_STEP},

            {"in",      1, 0, OPT_INPUT},
            {"out",     1, 0, OPT_OUTPUT},

            {0, 0, 0, 0}

        } ;
        c = getopt_long(argc,
                        argv,
                        "Lhm:i:o:b:e:x:f:s:",
                        long_options,
                        &option_index) ;
        if (c==-1) break ;

        switch(c) {

            /* Standard option: display license undocumented option */
            case OPT_LICENSE:
            case 'L':
                eclipse_display_license() ;
                return 0 ;
            /* Standard option : help */
            case OPT_HELP:
            case 'h':
                usage(argv[0]) ;
                break ;
            /* Standard option: version */
            case OPT_VERSION:
                print_eclipse_version() ;
                return 0 ;

            /* Local options: extraction modes  */
			case 'm':
	    		if (!strcmp(optarg, "plane"))
		    		emode = EXT_PLANE ;
		    	else if (!strcmp(optarg, "cube"))
			    	emode = EXT_CUBE ;
    			else if (!strcmp(optarg, "quad"))
	    			emode = EXT_QUAD ;
		    	else if (!strcmp(optarg, "pattern"))
			    	emode = EXT_PATTERN ;
    			else if (!strcmp(optarg, "list"))
	    			emode = EXT_LIST ;
		    	else if (!strcmp(optarg, "rect"))
			    	emode = EXT_RECTANGLE ;
    			else {
	    			e_error("unrecognized extraction mode : %s", optarg) ;
		    		return -1 ;
		    	}
			    break ;
			case OPT_PLANE:
			    emode = EXT_PLANE ;
			    break ;
			case OPT_CUBE:
    			emode = EXT_CUBE ;
	    		break ;
			case OPT_QUADRANT:
		    	emode = EXT_QUAD ;
			    break ;
			case OPT_PATTERN:
    			emode = EXT_PATTERN ;
	    		break ;
			case OPT_LIST:
		    	emode = EXT_LIST ;
			    break ;
			case OPT_RECTANGLE:
    			emode = EXT_RECTANGLE ;
	    		break ;

		    /* Local options: input and output file names */
			case OPT_INPUT:
			case 'i':
			    strncpy(name_i, optarg, FILENAMESZ) ;
    			break ;
			case OPT_OUTPUT:
			case 'o':
	    		strncpy(name_o, get_rootname(optarg), FILENAMESZ) ;
		    	break ;
    		/* Processing options : begin and end planes	*/
			case OPT_BEGIN:
			case 'b':
	    		begin = (int)atoi(optarg) ;
		    	break ;
			case OPT_END:
		    case 'e':
			    end = (int)atoi(optarg) ;
    			break ;
	    	/* Processing option : extraction string	*/
			case OPT_EXTSTRING:
			case 'x':
		    	strncpy(extstring, optarg, MAX_IMAGE_NUMBER-1) ;
			    break ;
    		/* Processing option: number of images per cycle step	*/
			case OPT_STEP:
			case 's':
	    		step = (int)atoi(optarg) ;
		    	break ;
    		/* Processing option : list file name	*/
			case OPT_LISTNAME:
			case 'f':
	    		strncpy(listname, optarg, FILENAMESZ) ;
		    	break ;
            default:
                usage(argv[0]) ;
              break ;
        }
	}

	/* Initialize eclipse environment */
	eclipse_init();

	if (name_i[0] == 0) {
		e_error("no input file name provided, use the -i option");
		return -1 ;
	}

	if (name_o[0] == 0) {
		strncpy(name_o, get_basename(get_rootname(name_i)), FILENAMESZ);
	}
	
	e_comment(1, "input     : %s", name_i);
	e_comment(1, "output    : %s", name_o);

	cube_extract(name_i, name_o, emode, begin, end, listname, extstring, step);

	if (debug_active()) xmemory_status() ;
    return 0 ;

}

static void usage(char * pname)
{
	hello_world(pname, prog_desc) ; 
	printf("use: %s -i infile [-o outfile] [mode options]\n", pname) ;
	printf("use: %s --in infile [--out outfile] [mode options]\n", pname) ;
	printf(
		"modes and associated options are:\n"
		"\n");
	printf(
		"plane extraction:\n"
		"\textract -m plane [-b <#>] [-e <#>]\n"
		"\textract --plane [--begin <#>] [--end <#>]\n"
		"\n");
	printf(
		"cube extraction:\n"
		"\textract -m cube [-b <#>] [-e <#>]\n"
		"\textract --cube [--begin <#>] [--end <#>]\n"
		"\n");
	printf(
		"quadrant extraction:\n"
		"\textract -m quad [-x '1234']\n"
		"\textract --quadrant [--ext '1234']\n"
		"\n");
	printf(
		"pattern extraction:\n"
		"\textract -m pattern [-x '011010...'] [-s <step>]\n"
		"\textract --pattern [--ext '011010...'] [--step <step>]\n"
		"\n");
	printf(
		"list extraction:\n"
		"\textract -m list [-f file]\n"
		"\textract --list [--name file]\n"
		"\n");
	printf(
		"rectangle extraction:\n"
		"\textract -m rect [-x 'llx lly urx ury']\n"
		"\textract --rectangle [--ext 'llx lly urx ury']\n"
		"\n\n");
	printf(
		"\tAlways supply an input name through -i or --in\n"
		"\tDefault output name depends on the mode. see doc\n"
		"\n\n") ;
	exit(0) ;
}

static int cube_extract(
        char			*	name_i,
    	char			*	name_o,
	    extraction_mode		emode,
        int					begin,
        int					end,
        char			*	listname,
        char			*	extstring,
        int					step)
{
    cube_t	*	in ;
	cube_t	*	ext ;
	cube_t	*	sav ;
	int			p ;
	char		cname_o[FILENAMESZ];
	char		cmt[ASCIILINESZ];
	FILE	*	listfile ;
	int			pnum ;
	int		*	ext_flags ;
	int			plen, ppos ;
	int			qnum ;
	int			llx, lly, urx, ury ;
	history	*	hs ;

	switch (emode) {
		/*
		 * ---------- Plane extraction
		 */ 
		case EXT_PLANE:
		e_comment(1, "extracting: planes");
		if (begin<0) {
			e_comment(1, "begin     : first");
		} else {
			e_comment(1, "begin     : %d", begin);
		}
		if (end<0) {
			e_comment(1, "end       : last");
		} else {
			e_comment(1, "end       : %d", end);
		}

		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}

		if (begin<0) begin=1 ;
		if (end<0)   end = in->np ;

		if ((begin<1) || (end>in->np) || (end<begin)) {
			e_error("begin/end incorrectly defined");
			cube_del(in);
			return -1 ;
		}
		/* Save planes one by one */
		for (p=begin ; p<=end ; p++) {
			sav = cube_from_image(in->plane[p-1]);
			if (sav==NULL) {
				e_error("cannot get plane %d from cube", p);
				cube_del(in);
				return -1 ;
			}
			hs = history_new() ;
			history_add(hs, "--- eclipse extract");
			history_add(hs, "input file:");
			history_add(hs, get_basename(name_i));
			history_add(hs, "this file is plane %d out of %d",p,in->np);

			sprintf(cname_o, "%s_%04d.fits", name_o, p);
			e_comment(1, "saving plane %d as [%s]", p, cname_o);
			cube_save_fits_hdrcopy_wh(sav, cname_o, name_i, hs) ;
			history_del(hs);
			cube_del(sav);
		}
		cube_del(in);
		break ;

		case EXT_CUBE:
		/*
		 * ---------- Cube extraction
		 */ 
		e_comment(1, "extracting: cube");
		if (begin<0) {
			e_comment(1, "begin     : first");
		} else {
			e_comment(1, "begin     : %d", begin);
		}
		if (end<0) {
			e_comment(1, "end       : last");
		} else {
			e_comment(1, "end       : %d", end);
		}
		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}

		if (begin<0) begin=1 ;
		if (end<0)   end = in->np ;

		if ((begin<1) || (end>in->np) || (end<begin)) {
			e_error("begin/end incorrectly defined");
			cube_del(in);
			return -1 ;
		}
		/* Reshuffle plane pointers */
		ext = cube_new(in->lx, in->ly, (end-begin+1));
		for (p=begin ; p<=end ; p++) {
			ext->plane[p-begin] = in->plane[p-1];
		}
		cube_del_shallow(in);
		/* Save result cube */
		hs = history_new() ;
		history_add(hs, "--- eclipse extract");
		history_add(hs, "input file:");
		history_add(hs, get_basename(name_i));
		history_add(hs, "planes %d to %d (incl)", begin, end);

		sprintf(cname_o, "%s.fits", name_o);
		e_comment(1, "saving cube [%d-%d] as [%s]", begin, end, cname_o);
		cube_save_fits_hdrcopy_wh(ext, cname_o, name_i, hs) ;
		history_del(hs);
		cube_del(ext);
		break ;

		case EXT_LIST:
		/*
		 * ---------- List extraction
		 */ 
		e_comment(1, "extracting: planes from a given list");
		e_comment(1, "listname  : %s", listname);
		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}
		if ((listfile=fopen(listname, "r"))==NULL) {
			e_error("cannot open list file [%s]", listname);
			cube_del(in);
			return -1 ;
		}
		ext_flags = calloc(in->np, sizeof(int));
		pnum = 0 ;
		while (fscanf(listfile, "%d", &p) != EOF) {
			ext_flags[p-1] = 1 ;
			pnum++ ;
		}
		fclose(listfile);
		ext = cube_new(in->lx, in->ly, pnum);
		hs = history_new();
		history_add(hs, "--- eclipse extract");
		history_add(hs, "input file:");
		history_add(hs, get_basename(name_i));
		pnum = 0 ;
		for (p=0 ; p<in->np ; p++) {
			if (ext_flags[p]) {
				ext->plane[pnum++] = in->plane[p] ;
				history_add(hs, "extracted plane %d out of %d",p+1,in->np);
			} else {
				image_del(in->plane[p]);
			}
			in->plane[p] = NULL ;
		}
		free(ext_flags);
		cube_del_shallow(in);
		sprintf(cname_o, "%s.fits", name_o);
		cube_save_fits_hdrcopy_wh(ext, cname_o, name_i, hs) ;
		history_del(hs);
		cube_del(ext);
		break ;

		case EXT_PATTERN:
		/*
		 * ---------- Pattern extraction
		 */ 
		e_comment(1, "extracting: cube with pattern");
		e_comment(1, "pattern   : %s", extstring);
		e_comment(1, "cyclestep : %d", step);
		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}
		/* Build extraction flag list */
		plen = strlen(extstring);
		if (plen<1) {
			e_error("invalid extraction string");
			cube_del(in);
			return -1 ;
		}
		ext_flags = calloc(in->np, sizeof(int));
		for (p=0 ; p<in->np ; p++) {
			ppos = (p/step) % plen ;
			if (extstring[ppos]=='1') {
				ext_flags[p] = 1 ;
			} else if (extstring[ppos]=='0') {
				ext_flags[p] = 0 ;
			} else {
				e_error("invalid character in pattern: [%s]", extstring);
				free(ext_flags);
				cube_del(in);
				return -1 ;
			}
		}
		for (p=0 ; p<in->np ; p++) {
			if (ext_flags[p]) {
				sav = cube_from_image(in->plane[p]);
				if (sav==NULL) {
					e_error("cannot get plane %d from cube", p+1);
					cube_del(in);
					free(ext_flags);
					return -1 ;
				}
				hs = history_new() ;
				history_add(hs, "--- eclipse extract");
				history_add(hs, "input file:");
				history_add(hs, get_basename(name_i));
				history_add(hs, "this file is plane %d out of %d",p+1,in->np);
				sprintf(cname_o, "%s_%04d.fits", name_o, p+1);
				e_comment(1, "saving plane %d as [%s]", p+1, cname_o);
				cube_save_fits_hdrcopy_wh(sav, cname_o, name_i, hs) ;
				history_del(hs);
				cube_del(sav);
			}
		}
		cube_del(in);
		free(ext_flags);
		break ;

		case EXT_QUAD:
		/*
		 * ---------- Quadrant extraction
		 */ 
		e_comment(1, "extracting: quadrants");
		e_comment(1, "quadrants : %s", extstring);
		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}
		plen = strlen(extstring);
		for (p=0 ; p<plen ; p++) {
			switch(extstring[p]) {
				case '1':
				ext = cube_getvig(in,
											 1,       		1+in->ly/2,
											 in->lx/2,  	in->ly);
				sprintf(cmt, "this is the upper left quadrant");
				qnum = 1 ;
				break ;

				case '2':
				ext = cube_getvig(in,
											 1+in->lx/2,  	1+in->ly/2,
											 in->lx,      	in->ly);
				sprintf(cmt, "this is the upper right quadrant");
				qnum = 2 ;
				break ;

				case '3':
				ext = cube_getvig(in,
											 1+in->lx/2,  	1,
											 in->lx,      	in->ly/2);
				sprintf(cmt, "this is the lower right quadrant");
				qnum = 3 ;
				break ;

				case '4':
				ext = cube_getvig(in,
											 1,         	1,
											 in->lx/2,  	in->ly/2);
				sprintf(cmt, "this is the lower left quadrant");
				qnum = 4 ;
				break ;

				default:
				e_error("unexpected quadrant ID in list [%s]", extstring);
				cube_del(in);
				return -1 ;
			}
			hs = history_new();
			history_add(hs, "--- eclipse extract");
			history_add(hs, "input file:");
			history_add(hs, get_basename(name_i));
			sprintf(cname_o, "%s_quad%d.fits", name_o, qnum);
			e_comment(1, "saving quadrant %d as [%s]", qnum, cname_o);
			cube_save_fits_hdrcopy_wh(ext, cname_o, name_i, hs) ;
			history_del(hs);
			cube_del(ext);
		}
		cube_del(in);
		break ;

		case EXT_RECTANGLE:
		/*
		 * ---------- Rectangle extraction
		 */ 
		e_comment(1, "extracting: rectangle");
		e_comment(1, "rectangle : %s", extstring);
		/* Load input cube */
		if ((in=cube_load(name_i))==NULL) {
			e_error("cannot load cube [%s]: aborting", name_i);
			return -1 ;
		}
		sscanf(extstring, "%d %d %d %d", &llx, &lly, &urx, &ury);
		ext = cube_getvig(in, llx, lly, urx, ury);
		cube_del(in);

		hs = history_new();
		history_add(hs, "--- eclipse extract");
		history_add(hs, "input file:");
		history_add(hs, get_basename(name_i));
		history_add(hs, "extraction zone: %d %d %d %d", llx, lly, urx, ury);

		sprintf(cname_o, "%s_ext.fits", name_o);
		e_comment(1, "saving extracted rectangle as [%s]", cname_o);
		cube_save_fits_hdrcopy_wh(ext, cname_o, name_i, hs) ;
		cube_del(ext);
		break ;

		case EXT_UNDEFINED:
		default:
		e_error("undefined extraction mode: aborting");
		return -1 ;
	}
	return 0 ;
}

